src/util/byteswap.c

```diff
@@ -59,13 +59,36 @@ uint64_t swap_uint64(uint64_t val)
              (val & 0x00FF000000000000UL) >> 40 | (val & 0xFF00000000000000UL) >> 56;
 }
 
-uint64_t swap_uint_length(uint64_t val, size_t bit_length)
+uint64_t decode_little_endian_bitmode(uint64_t val, size_t bit_length)
+/*
+ * Convert a variable bit length little endian encoded stream, accounting for byte-level
+ * endian granularity, also accounting for remaining bits
+ *
+ * example conversion:
+ * incoming 12 bit stream:  LLLLLLLL HHHHxxxx
+ * converts to: xxxxHHHH LLLLLLLL
+ *
+ * @param val incoming value
+ * @param bit_length number of bits to swap
+ * @return byte swapped value
+ */
 {
-    if (bit_length <= 8) return val;
-    if (bit_length <= 16) return swap_uint16(val);
-    if (bit_length <= 24) return swap_uint24(val);
-    if (bit_length <= 32) return swap_uint32(val);
-    if (bit_length <= 64) return swap_uint64(val);
-    panic(PANIC_CAUSE_UNREACHABLE);
-    return 0; /* to make compiler happy */
+        if (bit_length <= 8) return val;
+        if (bit_length <= 16){
```

Comment:
I think this can be simpler.  Try this:
```c
const size_t mask = (1 << bit_length) -1;
 if (bit_length <= 8) return val & mask;
 if (bit_length <= 16) return swap_uint16(val) & mask;
 if (bit_length <= 24) return swap_uint24(val) & mask;
 if (bit_length <= 32) return swap_uint32(val) & mask;
 panic(PANIC_CAUSE_UNREACHABLE);
return 0; /* to make compiler happy */
```

Reply:
If we were smart we could merge 4 methods into 1 by merging all the `swap_uintXX` methods and parameterizing it.

Reply:
I don't think this will work per the requirements. See the code comments for the function. 

We need to move the high bits to the right so we get this effect:

incoming 12 bit stream:  LLLLLLLL HHHHxxxx
converts to:   xxxxHHHH LLLLLLLL

If we simply swap as you describe, we'll have this:

incoming 12 bit stream:  LLLLLLLL HHHHxxxx
converts to: HHHHxxxx LLLLLLLL

Reply:
That is strange if that is how they are doing it because that is not how endianness works.  Endianness describes _byte_ ordering; _bit_ ordering never changes between machines (least significant bit is always on the right).  So accessing 12 bits of memory in little endian it would look like this: L7 L6 L5 L4 L3 L2 L1 L0   xxxxH3 H2 H1 H0.  If they are doing it that way that probably needs to be put in a different method since not everyone will do it that way @brentpicasso .  We can discuss if you need.

https://en.wikipedia.org/wiki/Endianness
https://www.webopedia.com/TERM/B/big_endian.html

Reply:
Right. it's been a bit of a brain bender to figure out, but since these bits are transmitted in a stream, the 12 bits are packed together without the leading blank bits  (x x x x H3 H2 H1 H0)

Can chat more in slack.